File: .claude/hooks/stop-dead-image-checker.mjs
#!/usr/bin/env node
// @hookstack stop-dead-image-checker
// Vérifie les images relatives cassées dans tous les fichiers Markdown du repo (Stop).
// Scan complet — couvre la dette existante, pas seulement les fichiers de la session.
// Gère les chemins relatifs ET les chemins absolus (résolus depuis public/).
// Purement Node.js (fs + path), sans réseau ni dépendance externe.
import { readFileSync, existsSync, readdirSync } from 'node:fs';
import { join, dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const SKIP_DIRS = new Set(['node_modules', '.git', 'dist', 'build', '.next', 'out', '.claude']);
// Capture ![alt](src) — uniquement les images (le ! est obligatoire)
const IMAGE_RE = /!\[([^\]]*)\]\(([^)]+)\)/g;

function stripCode(content) {
  // Supprime les blocs de code clôturés (``` ou ~~~) — multiline
  content = content.replace(/^```[\s\S]*?^```\s*$/gm, '');
  content = content.replace(/^~~~[\s\S]*?^~~~\s*$/gm, '');
  // Supprime les spans de code inline
  content = content.replace(/`[^`\n]+`/g, '``');
  return content;
}

function isExternal(src) {
  return src.startsWith('http') || src.startsWith('data:') || src.startsWith('//');
}

function walkMd(dir, { readdir = readdirSync, exists = existsSync } = {}) {
  if (!exists(dir)) return [];
  const results = [];
  for (const entry of readdir(dir, { withFileTypes: true })) {
    if (SKIP_DIRS.has(entry.name)) continue;
    const full = join(dir, entry.name);
    if (entry.isDirectory()) results.push(...walkMd(full, { readdir, exists }));
    else if (/\.mdx?$/.test(entry.name)) results.push(full);
  }
  return results;
}

export function run(_input, {
  projectDir = process.env.CLAUDE_PROJECT_DIR ?? process.cwd(),
  readFile = readFileSync,
  exists = existsSync,
  readdir = readdirSync,
} = {}) {
  const mdFiles = walkMd(projectDir, { readdir, exists });
  if (!mdFiles.length) return null;

  const broken = [];
  for (const file of mdFiles) {
    let content;
    try { content = readFile(file, 'utf8'); } catch { continue; }

    for (const [, , src] of stripCode(content).matchAll(IMAGE_RE)) {
      if (isExternal(src)) continue;

      let abs;
      if (src.startsWith('/')) {
        // Chemin absolu → résolu depuis public/ (convention Next.js et sites statiques)
        abs = join(projectDir, 'public', src);
      } else {
        abs = resolve(dirname(file), src);
      }

      if (!exists(abs)) {
        broken.push(`${file.replace(`${projectDir}/`, '')}  →  ${src}`);
      }
    }
  }

  if (!broken.length) return null;

  return {
    message:
      `[dead-image-checker] ${broken.length} broken image reference(s) across docs:\n` +
      broken.map((b) => `  - ${b}`).join('\n') +
      '\n',
  };
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stderr.write(JSON.stringify(result));
}

File: .claude/hooks/stop-dead-image-checker.py
#!/usr/bin/env python3
# @hookstack stop-dead-image-checker
"""Checks broken relative images across all Markdown files in the repo (Stop)."""
import json
import os
import re
import sys

SKIP_DIRS = {
    "node_modules",
    ".git",
    "dist",
    "build",
    ".next",
    "out",
    ".claude",
}
# Capture ![alt](src) — uniquement les images (le ! est obligatoire)
IMAGE_RE = re.compile(r"!\[([^\]]*)\]\(([^)]+)\)")


def _read_file(path):
    with open(path, "r", encoding="utf8") as f:
        return f.read()


def _exists(path):
    return os.path.exists(path)


def _readdir(path):
    with os.scandir(path) as it:
        return list(it)


def strip_code(content):
    # Supprime les blocs de code clôturés (``` ou ~~~) — multiline
    content = re.sub(r"^```[\s\S]*?^```\s*$", "", content, flags=re.M)
    content = re.sub(r"^~~~[\s\S]*?^~~~\s*$", "", content, flags=re.M)
    # Supprime les spans de code inline
    content = re.sub(r"`[^`\n]+`", "``", content)
    return content


def is_external(src):
    return src.startswith("http") or src.startswith("data:") or src.startswith("//")


def walk_md(directory, *, readdir, exists):
    if not exists(directory):
        return []
    results = []
    try:
        entries = readdir(directory)
    except Exception:
        return results
    for entry in entries:
        if entry.name in SKIP_DIRS:
            continue
        full = os.path.join(directory, entry.name)
        if entry.is_dir():
            results.extend(walk_md(full, readdir=readdir, exists=exists))
        elif re.search(r"\.mdx?$", entry.name):
            results.append(full)
    return results


def run(_input=None, *, project_dir=None, read_file=None, exists=None, readdir=None):
    if read_file is None:
        read_file = _read_file
    if exists is None:
        exists = _exists
    if readdir is None:
        readdir = _readdir
    if project_dir is None:
        project_dir = os.environ.get("CLAUDE_PROJECT_DIR", os.getcwd())

    md_files = walk_md(project_dir, readdir=readdir, exists=exists)
    if not md_files:
        return None

    broken = []
    for file in md_files:
        try:
            content = read_file(file)
        except Exception:
            continue

        for m in IMAGE_RE.finditer(strip_code(content)):
            src = m.group(2)
            if is_external(src):
                continue

            if src.startswith("/"):
                # Chemin absolu → résolu depuis public/ (convention Next.js et sites statiques)
                # (concaténation type path.join Node : un composant absolu ne réinitialise pas)
                abs_path = os.path.normpath(project_dir + os.sep + "public" + src)
            else:
                abs_path = os.path.normpath(os.path.join(os.path.dirname(file), src))

            if not exists(abs_path):
                broken.append(f"{file.replace(project_dir + '/', '')}  →  {src}")

    if not broken:
        return None

    return {
        "message": (
            f"[dead-image-checker] {len(broken)} broken image reference(s) across docs:\n"
            + "\n".join(f"  - {b}" for b in broken)
            + "\n"
        )
    }


if __name__ == "__main__":
    data = json.load(sys.stdin) if not sys.stdin.isatty() else None
    result = run(data)
    if result:
        sys.stderr.write(json.dumps(result))
